pkg/applier/kpt_applier.go

```diff
@@ -508,55 +528,79 @@ func (a *Applier) sync(ctx context.Context, objs []client.Object) (map[schema.Gr
-		klog.V(4).Infof("all resources are up to date.")
+	if a.Errors() == nil {
+		klog.V(4).Infof("Applier completed without error")
 	}
 
 	if stats.empty() {
-		klog.V(4).Infof("The applier made no new progress")
+		klog.V(4).Infof("Applier made no new progress")
 	} else {
-		klog.Infof("The applier made new progress: %s.", stats.string())
+		klog.Infof("Applier made new progress: %s.", stats.string())
 		objStatusMap.Log(klog.V(0))
 	}
-	return gvks, a.errs
+	return gvks
 }
 
-// Errors implements Interface.
 // Errors returns the errors encountered during apply.
+// Errors implements Interface.
 func (a *Applier) Errors() status.MultiError {
+	a.errsLock.RLock()
+	defer a.errsLock.RUnlock()
 	return a.errs
 }
 
-// Syncing implements Interface.
+func (a *Applier) addError(err error) {
+	// Ignore nil errors, to allow the caller to skip the nil check
+	if err == nil {
+		return
+	}
+	a.errsLock.Lock()
+	defer a.errsLock.Unlock()
+
+	// Default to an applier.Error
+	_, isStatusErr := err.(status.Error)
+	_, isMultiErr := err.(status.MultiError)
+	if !isStatusErr && !isMultiErr {
+		err = Error(err)
+	}
+	a.errs = status.Append(a.errs, err)
+}
+
+func (a *Applier) clearErrors() {
+	a.errsLock.Lock()
+	defer a.errsLock.Unlock()
+	a.errs = nil
+}
+
 // Syncing returns whether the applier is syncing.
+// Syncing implements Interface.
 func (a *Applier) Syncing() bool {
+	a.syncingLock.RLock()
+	defer a.syncingLock.RUnlock()
 	return a.syncing
 }
 
+func (a *Applier) setSyncing(syncing bool) {
+	a.syncingLock.Lock()
+	defer a.syncingLock.Unlock()
+	a.syncing = syncing
+}
+
+// Apply all managed resource objects and return any errors.
 // Apply implements Interface.
 func (a *Applier) Apply(ctx context.Context, desiredResource []client.Object) (map[schema.GroupVersionKind]struct{}, status.MultiError) {
-	// Clear the `errs` field at the start.
-	a.errs = nil
-	// Set the `syncing` field to `true` at the start.
-	a.syncing = true
-
-	defer func() {
-		// Make sure to clear the `syncing` field before `Apply` returns.
-		a.syncing = false
-	}()
-
-	a.mux.Lock()
-	defer a.mux.Unlock()
-
-	// Pull the actual resources from the API server to compare against the
-	// declared resources. Note that we do not immediately return on error here
-	// because the Applier needs to try to do as much work as it can on each
-	// cycle. We collect and return all errors at the end. Some of those errors
-	// are transient and resolve in future cycles based on partial work completed
-	// in a previous cycle (eg ignore an error about a CR so that we can apply the
-	// CRD, then a future cycle is able to apply the CR).
-	// TODO: Here and elsewhere, pass the MultiError as a parameter.
-	return a.sync(ctx, desiredResource)
+	a.syncLock.Lock()
+	defer a.syncLock.Unlock()
+
+	a.setSyncing(true)
+	defer a.setSyncing(false)
+
+	// Clear the errors at the start of a new apply.
+	// TODO: Keep track of old errors and new errors and merge them, until applyInner returns.
```

Comment:
It turns out the parse already does a DeepEquals on the before and after MultiError to prevent unnecessary updates, but when the async status updates were added, they stomp all over the cache and the invalidation done by the parser...

This all needs a rewrite, but I don't have the time or energy to do it now.